zero/client_server/client.py

```diff
@@ -159,7 +162,7 @@ async def try_connect(self):
         frames = [unique_id(), "connect", ""]
         await self.zmq.send(self._encoder.encode(frames))
         self._encoder.decode(await self.zmq.recv())
-        logging.info(f"Connected to server at {self._address}")
+        logging.info("Connected to server at %s", self._address)
```

Comment:
f string is faster?

Reply:
I believe it isn't.
Let's take a look at the following snippet I've written.
```py
import timeit

print(
    "Logging with f-string interpolation:",
    f_string_result := timeit.repeat(
        'logger.info(f"Connected to server at {address}")',
        setup='import logging; logger = logging.getLogger("test"); address = "tcp://localhost:5555"',
    ),
    f"(average: {sum(f_string_result) / len(f_string_result)})"
)

print(
    "Logging with % lazy formatting:",
    lazy_percent_result := timeit.repeat(
        'logger.info("Connected to server at %s", address)',
        setup='import logging; logger = logging.getLogger("test"); address = "tcp://localhost:5555"',
    ),
    f"(average: {sum(lazy_percent_result)/len(lazy_percent_result)})"
)
```
What are your results?
Mine are:
```
Logging with f-string interpolation: [0.5666271, 0.506797, 0.5651124999999999, 0.4688949, 0.5243291000000001] (average: 0.52635212)
Logging with % lazy formatting: [0.45181320000000014, 0.49188149999999986, 0.4773721000000002, 0.49471240000000005, 0.43005150000000025] (average: 0.4691661400000001)
```
on a Windows 10 machine,
```
Logging with f-string interpolation: [1.4035472849991493, 1.472289606999766, 1.3470586610001192, 1.5006198740011314, 1.523844381001254] (average: 1.449471961600284)
Logging with % lazy formatting: [1.2818151179999404, 1.2488627920010913, 1.4321971019999182, 1.3152025340004911, 1.2156276049990993] (average: 1.2987410302001081)
```
on a Linux-5.15 machine, and
```
Logging with f-string interpolation: [1.428359266370535, 1.0918207131326199, 1.121897578239441, 1.1788691692054272, 1.4921034537255764] (average: 1.26261003613472)                                                                             Logging with % lazy formatting: [1.5025393292307854, 0.9335274249315262, 0.9860474318265915, 1.579454306513071, 1.0524444356560707] (average: 1.210802585631609)
```
on a FreeBSD-13.0 machine.
What we can infer from the above is that no matter the platform the %-style formatting tends to be average faster than the f-string-based one.

Reply:
Interesting 🤔 I read about f-string are faster [here](https://www.reddit.com/r/Python/comments/pivojb/performance_comparison_of_string_formatting/)

The result of your benchmark on my machine (windows 11) is -
```
Logging with f-string interpolation: [0.2028639000000112, 0.21008989999972982, 0.21050130000003264, 0.20816390000072715, 0.21055859999978566] (average: 0.2084355200000573)
Logging with % lazy formatting: [0.18073139999978594, 0.1822362000002613, 0.18509260000064387, 0.17878840000048513, 0.1782229999989795] (average: 0.18101432000003115)
```

Seems like % is faster.

Then we should be consistent in the whole codebase right? 🤔

Reply:
What you read is 100% correct.
It's not really about %-formatting being faster than f-string formatting _in general_.
It's false. F-strings are always faster. Using f-strings is also always better, _if it's really useful_.
When it comes to logging, it's about laziness. Behind the scenes, if we use the lazy-style formatting (which, in logging, happens by default with % as in my change, but you can also do something like [this](https://github.com/pylint-dev/pylint/blob/cb255eaaed8bba6bec1f7bf5d4cde15821c1dd46/doc/data/messages/l/logging-not-lazy/bad.py#L6) and it won't be lazy), it will be evaluated once a log is actually emitted. In the current code, the log records are formatted no matter whether they will get emitted or not. In the snippet I've sent, the logs were not really emitted to any stream.
That's why it is advised to use lazy formatting, because in most cases it saves time. Some loggers may have more handlers than the other ones, etc., etc., so it's not a trivial case.

To prove that what you've read was correct, let's run this:

```py
import timeit

setup_stmt = (
    'import logging; import os; logger = logging.getLogger("test"); '
    'logger.setLevel(logging.INFO); '
    'logger.addHandler(logging.StreamHandler(open(os.devnull, "w"))); '
    'address = "tcp://localhost:5555"'
)

print(
    "Logging with f-string interpolation:",
    timeit.timeit(
        'logger.info(f"Connected to server at {address}")',
        setup=setup_stmt, number=100000
    ),
)

print(
    "Logging with % lazy formatting:",
    timeit.timeit(
        'logger.info("Connected to server at %s", address)',
        setup=setup_stmt, number=100000
    ),
)
```
the `setup_stmt` above ensures that the logger will _actually_ have to format the message.
And f-string solution is faster now, just because the log formatting _actually_ occurs.

My results:
```
Logging with f-string interpolation: 2.5147972000000003
Logging with % lazy formatting: 4.2350709
```
(Windows 10 machine)

This 'tiny inconsistence' should not be considered inconsistence. It's about using proper tools for proper use cases.

Reply:
Amazing! I used lazy logging evaluation in kotlin but didn't know about this in python. Thanks a lot for the good examples! 🙏

And yes it's better to use "proper tools for the proper use cases" 🙌